### ISSUE TITLE
Max message size validation should be at batch level and not at message level

### ISSUE THREAD

**Issue by kishaningithub** [segmentio/kafka-go#720]
**Describe the bug**

The behavior of max.message.bytes has changed in the recent versions of kafka

> The broker configuration max.message.bytes now applies to the total size of a batch of messages. Previously the setting applied to batches of compressed messages, or to non-compressed messages individually. A message batch may consist of only a single message, so in most cases, the limitation on the size of individual messages is only reduced by the overhead of the batch format. However, there are some subtle implications for message format conversion (see below for more detail). Note also that while previously the broker would ensure that at least one message is returned in each fetch request (regardless of the total and partition-level fetch sizes), the same behavior now applies to one message batch.

https://kafka.apache.org/26/documentation.html

**Kafka Version**
2.6.x

**To Reproduce**

- Ensure `max.message.bytes` is 1MB in kafka
- Configure a writer with `BatchSize` as  `1000`
- Writer 1000 messages each having size 2KB (Point is 1 batch should exceed 1MB)

**Expected behavior**

Below is my thought on what must happen
- The current batch must be trimmed to max `BatchBytes` and a new batch must be created and no error must be returned to the consumer of this library.
- If the trimming above cannot happen because a single message exceeds `BatchBytes` then `messageTooLarge` error must be returned

**Actual behavior**
- The batch passes through the kafka-go library as the check there is done at a message level and not at a batch level. Due to this the batch fails with Write error from kafka

**Additional context**

https://github.com/segmentio/kafka-go/blob/05c9b5b01ae26403314df306c4e3d228d5e5f7be/writer.go#L586

If you observe this line the check is done at a message level and not at a batch level.

**Comment by rhansen2** [comment]
Hi @kishaningithub, thanks for submitting this issue.

I'm having a bit of trouble reproducing the behavior you describe in the `Actual Behavior` section. I'm seeing the messages get split into two batches based on the size of the messages exceeding `BatchBytes`.

Perhaps there's something wrong with my setup.

I'm running kafka via docker with the following compose file:
```
version: "3"
services:
  kafka:
    image: wurstmeister/kafka:2.13-2.6.0
    restart: on-failure:3
    links:
    - zookeeper
    ports:
    - 9092:9092
    - 9093:9093
    environment:
      KAFKA_VERSION: '2.6.0'
      KAFKA_BROKER_ID: '1'
      KAFKA_CREATE_TOPICS: 'test-writer-0:1:1'
      KAFKA_DELETE_TOPIC_ENABLE: 'true'
      KAFKA_MAX_MESSAGE_BYTES: '1048576'
      KAFKA_ADVERTISED_HOST_NAME: 'localhost'
      KAFKA_ADVERTISED_PORT: '9092'
      KAFKA_ZOOKEEPER_CONNECT: 'zookeeper:2181'
      KAFKA_AUTO_CREATE_TOPICS_ENABLE: 'true'
      KAFKA_LISTENERS: 'PLAINTEXT://:9092'
      KAFKA_ADVERTISED_LISTENERS: 'PLAINTEXT://localhost:9092'

  zookeeper:
    image: wurstmeister/zookeeper
    ports:
    - 2181:2181
```
and the sample program I'm using is
```go
package main

import (
	"bytes"
	"context"
	"fmt"

	"github.com/segmentio/kafka-go"
)

type logger struct{}

func (l logger) Printf(s string, args ...interface{}) {
	fmt.Printf(s+"\n", args...)
}

func main() {
	w := &kafka.Writer{
		Addr:         kafka.TCP("localhost:9092"),
		Topic:        "test-writer-0",
		BatchSize:    1000,
		Logger:       logger{},
		RequiredAcks: kafka.RequireAll,
	}
	defer w.Close()

	message := make([]kafka.Message, 1000)
	for i := 0; i < len(message); i++ {
		var val []byte
		val = bytes.Repeat([]byte{0}, 2048)
		message[i] = kafka.Message{
			Key:   []byte(fmt.Sprintf("Key%d", i)),
			Value: val,
		}
	}
	if err := w.WriteMessages(context.Background(), message...); err != nil {
		panic(err)
	}
}

```
Please let me know if there's anything wrong with my reproduction or if I'm misunderstanding the issue.

Thanks!

**Comment by kishaningithub** [comment]
@rhansen2 Thanks for the above setup code. When we tried with the above code, we figured the actual issue we were facing was due to a mismatch between the `BatchBytes` configuration in this library and the one we had in our kafka instance `max.message.bytes`. 

In our case the `max.message.bytes` was 1 MB and the `BatchBytes` was 10 MB causing the library to send out a 10 MB batch while the kafka can only accept 1 MB

Closing this issue as these 2 values are expected to be in sync.
